src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,382 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+      if (isLeadSurrogate(c) && inputPos + 1 < input.size() &&
+          isTrailSurrogate(input[inputPos + 1])) {
+        // Valid surrogate pair = 4 bytes in UTF-8
+        utf8Length += 4;
+        inputPos += 2;
+      } else {
+        // Invalid surrogate = U+FFFD (3 bytes)
+        utf8Length += 3;
+        inputPos++;
+      }
+    } else {
+      // Unexpected error - fall back to scalar calculation for safety
+      break;
+    }
+  }
+
+  return utf8Length;
+}
+
+// Encode a single UTF-16 code unit to UTF-8
+inline size_t encodeUtf8CodeUnit(char16_t c, kj::ArrayPtr<char> out) {
+  if (c < 0x80) {
+    out[0] = static_cast<char>(c);
+    return 1;
+  } else if (c < 0x800) {
+    out[0] = static_cast<char>(0xC0 | (c >> 6));
+    out[1] = static_cast<char>(0x80 | (c & 0x3F));
+    return 2;
+  } else {
+    out[0] = static_cast<char>(0xE0 | (c >> 12));
+    out[1] = static_cast<char>(0x80 | ((c >> 6) & 0x3F));
+    out[2] = static_cast<char>(0x80 | (c & 0x3F));
+    return 3;
+  }
+}
+
+// Encode a valid surrogate pair to UTF-8
+inline void encodeSurrogatePair(char16_t lead, char16_t trail, kj::ArrayPtr<char> out) {
+  uint32_t codepoint = 0x10000 + (((lead & 0x3FF) << 10) | (trail & 0x3FF));
+  out[0] = static_cast<char>(0xF0 | (codepoint >> 18));
+  out[1] = static_cast<char>(0x80 | ((codepoint >> 12) & 0x3F));
+  out[2] = static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F));
+  out[3] = static_cast<char>(0x80 | (codepoint & 0x3F));
+}
+
+// Convert UTF-16 with potentially invalid surrogates to UTF-8.
+// Invalid surrogates are replaced with U+FFFD.
+// Returns the number of UTF-8 bytes written.
+// Uses SIMD for valid portions and falls back to scalar for invalid surrogates.
+size_t convertInvalidUtf16ToUtf8(kj::ArrayPtr<const char16_t> input, kj::ArrayPtr<char> out) {
+  size_t inputPos = 0;
+  size_t outputPos = 0;
+
+  while (inputPos < input.size()) {
+    // Find the next invalid surrogate using SIMD validation
+    auto result =
+        simdutf::validate_utf16_with_errors(input.begin() + inputPos, input.size() - inputPos);
+
+    if (result.error == simdutf::error_code::SUCCESS) {
+      // Remaining input is valid - convert it all with SIMD
+      outputPos += simdutf::convert_utf16_to_utf8(
+          input.begin() + inputPos, input.size() - inputPos, out.begin() + outputPos);
+      break;
+    }
+
+    if (result.error == simdutf::error_code::SURROGATE) {
+      // Convert the valid portion before the error with SIMD
+      if (result.count > 0) {
+        outputPos += simdutf::convert_valid_utf16_to_utf8(
+            input.begin() + inputPos, result.count, out.begin() + outputPos);
+        inputPos += result.count;
+      }
+
+      // Handle the invalid surrogate at inputPos
+      char16_t c = input[inputPos];
+      if (isLeadSurrogate(c) && inputPos + 1 < input.size() &&
+          isTrailSurrogate(input[inputPos + 1])) {
+        // Valid surrogate pair - encode it (this shouldn't happen if SURROGATE error)
```

Comment:
Curious... if the `result.error == simdutf::error_code::SURROGATE` is true, doesn't that mean this isn't a valid surrogate? Even the comment here acknowledges that this case shouldn't happen. Is this case covered as a Just In Case?

Reply:
I removed this branch and added an assertion.